crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -279,66 +295,343 @@ where
+                packet_data.token.denom.to_string(),
+                Some(&Address::Internal(InternalAddress::Ibc)),
+            )?;
+            let token = namada_ibc::received_ibc_token(
+                ibc_denom,
+                &msg.message.packet.port_id_on_a,
+                &msg.message.packet.chan_id_on_a,
+                &msg.message.packet.port_id_on_b,
+                &msg.message.packet.chan_id_on_b,
+            )
+            .into_storage_result()
+            .map_err(Error::NativeVpError)?;
+            let delta = ValueSum::from_pair(
+                token.clone(),
+                Amount::from_uint(Uint(*packet_data.token.amount), 0).unwrap(),
+            );
+            // Enable funds to be taken from the IBC internal
+            // address and be deposited elsewhere
+            // Required for the IBC internal Address to release
+            // funds
+            let pre_entry =
+                acc.pre.entry(addr_taddr(IBC)).or_insert(ValueSum::zero());
+            *pre_entry =
+                checked!(pre_entry + &delta).map_err(native_vp::Error::new)?;
+        }
+        Ok(())
+    }
+
+    // Apply relevant IBC packets to the changed balances structure
+    fn apply_ibc_packet(
+        &self,
+        mut acc: ChangedBalances,
+        ibc_msg: &IbcMessage,
+        keys_changed: &BTreeSet<Key>,
+        ibc_acks: &BTreeMap<Vec<u8>, AcknowledgementStatus>,
+    ) -> Result<ChangedBalances> {
+        match ibc_msg {
+            // This event is emitted on the sender
+            IbcMessage::Transfer(msg) => {
+                // Get the packet commitment from post-storage that corresponds
+                // to this event
+                let receiver = msg.message.packet_data.receiver.to_string();
+                let addr = TAddrData::Ibc(receiver.clone());
+                acc.decoder.insert(ibc_taddr(receiver), addr);
+                self.apply_send_packet(&mut acc, msg, keys_changed)?;
+            }
+            // This event is emitted on the receiver
+            IbcMessage::RecvPacket(msg) => {
+                // Check if this is a Transfer packet
+                let Ok(packet_data) = serde_json::from_slice::<PacketData>(
+                    &msg.message.packet.data,
+                ) else {
+                    return Ok(acc);
+                };
+                let receiver = packet_data.receiver.to_string();
+                let addr = TAddrData::Ibc(receiver.clone());
+                acc.decoder.insert(ibc_taddr(receiver), addr);
+                self.apply_write_ack(
+                    &mut acc,
+                    msg,
+                    &packet_data,
+                    keys_changed,
+                    ibc_acks,
+                )?;
+            }
+            // Ignore all other IBC events
+            _ => {}
+        }
+        Result::<_>::Ok(acc)
+    }
+
+    // Apply the balance change to the changed balances structure
+    fn apply_balance_change(
```

Comment:
Could all these functions simply be exported by the `ibc` crate? Also, could they be combined into one which takes all the state changes and returns the `ChangedBalances` for this transaction which the MASP VP needs?

Reply:
Moving the `ChangedBalances` functions into the `namada_ibc` crate is problematic because those functions depend on the `namada_vp_env` crate (which itself depends on the `namada_ibc` crate). The latter dependency is due to those functions doing pre and post balance reads as well as IBC trace reads and packet commitment reads from storage. I suspect I'd have to restructure these functions in order to reduce their dependence on validity predicate functions first. Should I attempt restructuring those functions in this PR?

Reply:
(discussed offline - let's do this separately)